cpp/examples/arrow/CMakeLists.txt

```diff
@@ -36,10 +36,14 @@ if(ARROW_FLIGHT)
   # we'll violate ODR for gRPC symbols
   if(ARROW_GRPC_USE_SHARED)
     set(FLIGHT_EXAMPLES_LINK_LIBS arrow_flight_shared)
-    # We don't directly use symbols from the reflection library, so
-    # ensure the linker still links to it
-    set(GRPC_REFLECTION_LINK_LIBS -Wl,--no-as-needed gRPC::grpc++_reflection
-                                  -Wl,--as-needed)
+    if(APPLE)
```

Comment:
Should we branch on APPLE or on Clang? (Does Clang support the option on other platforms?) e.g. 

```
if(CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang" OR CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
```

Reply:
Good point, let me check, I'll use docker for this

Reply:
It seems it is something with the Apple toolchain, because clang can build this example on ubuntu.  So, I think that branching on APPLE make sense here. I tested on:
```bash
Ubuntu clang version 14.0.0-1ubuntu1
Ubuntu 22.04.1 LTS
Linux 7122ee4a21c0 5.10.104-linuxkit #1 SMP PREEMPT Thu Mar 17 17:05:54 UTC 2022 aarch64 aarch64 aarch64 GNU/Linux
```
btw docker on macos m1 is kind of slow.